dataset/src/main/scala/frameless/TypedColumn.scala

```diff
@@ -188,8 +190,16 @@ sealed class TypedColumn[T, U](
     *
     * apache/spark
     */
-  def multiply(u: TypedColumn[T, U])(implicit n: CatalystNumeric[U]): TypedColumn[T, U] =
-    self.untyped.multiply(u.untyped).typed
+  def multiply(u: TypedColumn[T, U])(implicit n: CatalystNumeric[U], ct: ClassTag[U]): TypedColumn[T, U] = {
+    if (ct.runtimeClass == BigDecimal(0).getClass) {
+      // That's apparently the only way to get sound multiplication.
+      // See https://issues.apache.org/jira/browse/SPARK-22036
+      val dt = DecimalType(20, 14)
```

Comment:
This `DecimalType` (see https://spark.apache.org/docs/2.2.0/api/java/org/apache/spark/sql/types/DecimalType.html) was suggested by Marco Gaido in https://issues.apache.org/jira/browse/SPARK-22036, right? What makes this a good overall choice? Wouldn't it be better to do the cast in the test rather than here?

Reply:
This is the suggestion from SPARK-22036.

> What makes this a good overall choice?

I did some trial and error and this seems to be the maximum precision/scale under which multiplication never returns null.

> Wouldn't it be better to do the cast in the test rather than here?

If we don't do that systematically we would have to change the return type of `BigDecimal` multiplication to be `Option[BigDecimal]`. If SPARK-22036 is indeed a bug then I think it's better to have a workaround than to "encode the bug in the types". If it's "working as intended", then we should probably go do for `Option[BigDecimal]`...